### PR TITLE
Recovery Pallet benchmarking

### PR DESCRIPTION
**Pull request by ayevbeosa**
As raised in #11093, `benchmarking.rs` has been created for the `recovery-pallet` to produce and measure weights.

closes: https://github.com/paritytech/substrate/issues/11093

Polkadot address: 12dVgRw6ntsMQiETAoCBzu6MA7LPWagvU4Y1rH9jGPDHQvPX

polkadot companion: https://github.com/paritytech/polkadot/pull/5366

### CODE REVIEW

**Comment by ggwpez** [comment]
Thanks @ayevbeosa !  
If you are ready with the code then just click on the "Ready for review" button and I will take a look.

**Comment by ggwpez** [comment]
Sorry I did not yet have the the time to look into it, but the [CI reports](https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1484287) that some of the benches fail.  
You should be able to reproduce this with `cargo test --features=runtime-benchmarks -p pallet-recovery`.

**Comment by ayevbeosa** [comment]
Alright thanks

**Comment by ayevbeosa** [comment]
It seems having a large `MaxFriends` caused an existing test to fail. I reduced the number and edited the test to work with it if that is okay.

File: frame/recovery/src/benchmarking.rs [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+use crate::Pallet;
+use frame_benchmarking::{account, benchmarks, whitelisted_caller};
+use frame_support::traits::{Currency, Get};
+use frame_system::RawOrigin;
+use sp_runtime::traits::{BlockNumberProvider, Bounded};
+
+const SEED: u32 = 0;
+
+#[allow(dead_code)]
```
Comment (ggwpez) [inline frame/recovery/src/benchmarking.rs:30]:
```suggestion
```

File: frame/recovery/src/benchmarking.rs [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+		T::Currency::make_free_balance_be(
+			&friends.get(friend).unwrap(),
+			BalanceOf::<T>::max_value(),
+		);
+	}
+
+	friends
+}
+
+fn add_caller_and_generate_friends<T: Config>(
```
Comment (ggwpez) [inline frame/recovery/src/benchmarking.rs:70]: This function is mostly the same as the one above.

**Review by ggwpez — COMMENTED** [review]
Looking real good!  
In general you could maybe compact the diff, since it looks a bit repetitive sometimes.  

The next step would then to create a weight file and use the weights to annotate the extrinsics.  
You can already do that, for example by using this [weight.rs.txt](https://github.com/paritytech/substrate/files/8474613/weight.rs.txt) (or creating your own).  
Otherwise we can also do that in another MR, since the bench-bot does not push to external repos AFAIK and I dont want to run the benches manually.

File: frame/recovery/src/benchmarking.rs (hunk 2 above)
Comment (ayevbeosa) [reply frame/recovery/src/benchmarking.rs:69]: Alright, I will work on this

**Comment by ayevbeosa** [comment]
From the errors in the CI reports, does this mean I have make a companion PR?

**Review by ggwpez — APPROVED** [review]
I pushes some final fixes, should be good now.

>From the errors in the CI reports, does this mean I have make a companion PR?  

Yes looks like it since it is already deployed there.

File: frame/recovery/src/benchmarking.rs [hunk 3]
```diff
@@ -0,0 +1,384 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2022 Parity Technologies (UK) Ltd.
```
Comment (kianenigma) [inline frame/recovery/src/benchmarking.rs:3]:
```suggestion
// Copyright (C) 2022 Parity Technologies (UK) Ltd.
```

**Comment by ayevbeosa** [comment]
> I pushes some final fixes, should be good now.

Thanks!

File: frame/recovery/src/benchmarking.rs [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+		let caller: T::AccountId = whitelisted_caller();
+		let recovered_account: T::AccountId = account("recovered_account", 0, SEED);
+		let call: <T as Config>::Call = frame_system::Call::<T>::remark { remark: vec![] }.into();
+
+		Proxy::<T>::insert(&caller, &recovered_account);
+	}: _(
+		RawOrigin::Signed(caller),
+		recovered_account,
+		Box::new(call)
+	)
```
Comment (kianenigma) [inline frame/recovery/src/benchmarking.rs:123]: you could use `remark_with_event` here and assert it actually executed.

File: frame/recovery/src/benchmarking.rs [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+
+	T::ConfigDepositBase::get().checked_add(&friend_deposit)
+}
+
+fn generate_friends<T: Config>(num: u32) -> Vec<<T as frame_system::Config>::AccountId> {
+	// Create friends
+	let mut friends = vec![];
+	for i in 0..num {
+		friends.push(account("friend", i, SEED));
+	}
```
Comment (kianenigma) [inline frame/recovery/src/benchmarking.rs:54]:
```suggestion
	// Create friends
	let friends = (0..num).map(|x| account("friend", i, SEED)).collect::<Vec<_>>();
```

File: frame/recovery/src/benchmarking.rs [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+		let caller: T::AccountId = whitelisted_caller();
+		let lost_account: T::AccountId = account("lost_account", 0, SEED);
+		let rescuer_account: T::AccountId = account("rescuer_account", 0, SEED);
+
+		// Create friends
+		let friends = add_caller_and_generate_friends::<T>(caller.clone(), n);
+		let bounded_friends: FriendsOf<T> = friends.try_into().unwrap();
+
+		let delay_period = block_number::<T>();
```
Comment (kianenigma) [inline frame/recovery/src/benchmarking.rs:188]: this doesn't really make sense to me. The delay should be some fixed, non-zero amount. In realistic benchmarks, this value coming from `block_number` will be 0. Is that intentional?

File: frame/recovery/src/benchmarking.rs [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+		assert_last_event::<T>(
+			Event::RecoveryVouched {
+				lost_account: lost_account,
+				rescuer_account: rescuer_account,
+				sender: caller,
+			}.into()
+		);
+	}
+
+	claim_recovery {
```
Comment (kianenigma) [inline frame/recovery/src/benchmarking.rs:230]: not a big deal, but a lot of the code in `create_recovery`, `vouch_recovery`, `claim_recovery` are common, and build on top of one another. You can have something like 

```rust
enum RecoveryStage {
  Create,
  Vouche(u16),
  Claim,
}

fn create_recovery_scenario(stage: RecoveryStage) {}
```

**Review by kianenigma — CHANGES_REQUESTED** [review]
Looks mostly good, but I have some questions and suggestions that I want to see answered first.

**Comment by kianenigma** [comment]
tip submitted: https://polkadot.polkassembly.io/tip/0x104a0b13dcc86a4a7c393ea3f8a30965b04a259108618adac9ee3f7a15a6b7d4

File: frame/recovery/src/benchmarking.rs (hunk 4 above)
Comment (ayevbeosa) [reply frame/recovery/src/benchmarking.rs:123]: I am unfamiliar with this, but I will give it a try.

File: frame/recovery/src/benchmarking.rs (hunk 7 above)
Comment (ayevbeosa) [reply frame/recovery/src/benchmarking.rs:227]: Hmm...okay, I can play around with this and make things cleaner.

File: frame/recovery/src/benchmarking.rs (hunk 6 above)
Comment (ayevbeosa) [reply frame/recovery/src/benchmarking.rs:185]: Oh it wasn't, using `block_number` felt like a better option at the time, I will edit it to use a non-zero amount then.

File: frame/recovery/src/benchmarking.rs (hunk 4 above)
Comment (ayevbeosa) [reply frame/recovery/src/benchmarking.rs:120]: I am having issues getting this to work, I wouldn't mind some help.

Comment (kianenigma) [reply frame/recovery/src/benchmarking.rs:120]: do you have a commit of your attempt?

File: frame/recovery/src/benchmarking.rs [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+use frame_system::RawOrigin;
+use sp_runtime::traits::Bounded;
+
+const SEED: u32 = 0;
+
+fn assert_last_event<T: Config>(generic_event: <T as Config>::Event) {
+	frame_system::Pallet::<T>::assert_last_event(generic_event.into());
+}
+
+fn block_number<T: Config>() -> T::BlockNumber {
```
Comment (kianenigma) [inline frame/recovery/src/benchmarking.rs:34]: also, 

- why is this a function? 
- why is it generic over `T`?

**Review by kianenigma — APPROVED** [review]
If you fix this https://github.com/paritytech/substrate/pull/11176/files#r855149197, this would be "good enough", so won't block it any further. Thanks for your contribution!

File: frame/recovery/src/benchmarking.rs (hunk 4 above)
Comment (ayevbeosa) [reply frame/recovery/src/benchmarking.rs:120]: This is what I did

```
	as_recovered {
		let caller: T::AccountId = whitelisted_caller();
		let recovered_account: T::AccountId = account("recovered_account", 0, SEED);
		let call: <T as Config>::Call = frame_system::Call::<T>::remark_with_event { remark: vec![] }.into();
		let call_hash = <T as frame_system::Config>::Hashing::hash_of(&call);

		Proxy::<T>::insert(&caller, &recovered_account);
	}: _(
		RawOrigin::Signed(caller.clone()),
		recovered_account,
		Box::new(call)
	) verify {
		assert_last_event(
			frame_system::<T>::Event::Remarked { sender: caller, hash: call_hash }
		);
	}
```

File: frame/recovery/src/benchmarking.rs (hunk 8 above)
Comment (ayevbeosa) [reply frame/recovery/src/benchmarking.rs:34]: Oh, this makes sense.

**Comment by ggwpez** [comment]
bot merge

**Pull request merged**